pkg/chunk/aws/metrics_autoscaling.go

```diff
@@ -0,0 +1,218 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	promApi "github.com/prometheus/client_golang/api"
+	promV1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	"github.com/weaveworks/common/mtime"
+
+	"github.com/weaveworks/cortex/pkg/chunk"
+	"github.com/weaveworks/cortex/pkg/util"
+)
+
+const (
+	cachePromDataFor       = 30 * time.Second
+	queueObservationPeriod = 2 * time.Minute
+	targetScaledown        = 0.1 // consider scaling down if queue smaller than this times target
+	targetMax              = 10  // always scale up if queue bigger than this times target
+	errorFractionScaledown = 0.1
+	minUsageForScaledown   = 100 // only scale down if usage is > this DynamoDB units/sec
+)
+
+type metricsData struct {
+	queueLengthTarget int64
+	scaleUpFactor     float64
+	promAPI           promV1.API
+	promLastQuery     time.Time
+	tableLastUpdated  map[string]time.Time
+	queueLengths      []float64
+	errorRates        map[string]float64
+	usageRates        map[string]float64
+}
```

Comment:
There is quite a lot of coupling between `metricsData`  and `dynamoTableClient`; I wonder if more work needs to be done to tease out a clean interface?

Reply:
OK, I think much of this was simply the metrics methods were on the wrong struct, but I went ahead and created an abstract interface so the choice between old-style and new-style is in the constructor.  Let me know what you think.

Reply:
Looks good now!